docs/proposals/spec-reconciliation.md

```diff
@@ -0,0 +1,91 @@
+# Spec Reconciliation Proposal
+
+## Introduction
+Add a way for broker to sync the list of specs it has with registries while handling the following corner cases:
+ 1. Temporary loss of connection with registry
+ 2. A spec removed from the registry but has provisioned instance/s
+
+## Problem Description
+
+The current approach of syncing the spec from registries is as follows:
+ 1. Delete all the specs currently in the datastore
+ 2. Fetch new specs
+ 3. Add all the fetched specs to the datastore
+ 
+This leads to a bunch of bugs [Bug 1583495](https://bugzilla.redhat.com/show_bug.cgi?id=1583495), [Bug 1577810](https://bugzilla.redhat.com/show_bug.cgi?id=1583495), [Bug 1586345](https://bugzilla.redhat.com/show_bug.cgi?id=1586345)
+
+
+Instead, the broker can take the following actions:
+####Approach#1
+ 1. Fetch the specs from registry
+    1. If the fetch fails increment a failed connection counter. If the failed connection counter reaches a limit delete all the specs that do not have a service instance
+    2. If the fetch succeeds:
+        1. add new specs to the datastore
+        2. delete that specs that are removed from datastore and do not have a service instance.
+        3. mark the specs that are removed from datastore and has a service instance
+ 2. During deprovision, if the instance getting deprovisioned is the last one delete the spec from datastore
```

Comment:
How would you know this is the last service instance associated with a particular bundle spec?

Reply:
Either maintain a counter of the number of instances provisioned in bundle.Spec or compare all the instance specs with the spec of the instance being deprovisioned. I think the former would be better

Reply:
The bundle spec should only get deleted in the case that it had been previously marked for deletion, but was *not* deleted because it had an outstanding instance, correct?